### PR TITLE
Doc hard, and with a vengeance

### PR DESCRIPTION
**Pull request by MikeInnes**
Alright, so for some reason I've come back for round two. This time I hope that what's going on is a lot clearer, and it's more automated so that it'll be easier to wait for input from whomever wants to provide it.

Remaining issues:
- Quote escaping
- Module clashes – the helpdb.jl production treats `eps` as separate from `Dates.eps`, but they're actually the same function and so one clobbers the other.
- Multiple methods – these are all joined together for now. Technically we could fix this in a hacky way, but it would be fiddly – I think it's best to wait for a real solution with the new manual.
- Shared docstrings, e.g. `∉` and `∋` with `in`.

I think there are few enough of these cases (aside from the methods thing) that we can fix them by hand later.

From a quick check, a relatively small number (222) of the docstrings use weird `:...:` syntaxes which means the rest (1036) should be easy to convert to markdown automatically.

### CODE REVIEW

**Comment by shashi** [comment]
This looks lossless so far! :+1:

How do you want to handle RST specific features? Interpolating Julia code, maybe? e.g. `:func:f` could be `$(func_link(function_name)`. And similar functions for other things?

**Comment by jiahao** [comment]
Thank you for trying again. I'll give this more careful scrutiny tomorrow, but can you explain what the idea is in this current iteration? For example, are the lines following `.. function:: test(pkgs...)` meant to be the new docstring for `Pkg.test()`? https://github.com/JuliaLang/julia/commit/66a5b0f2545cb3aa23ccb7d9a44e5d17c8221d56#diff-bf56c1110568368fd8d56907a8df71d9R220

```rest
.. function:: test(pkgs...)
 
   ``test()``

   Run the tests for all installed packages ensuring that each package's test dependencies are installed for the duration of the test. A package is tested by running its ``test/runtests.jl`` file and test dependencies are specified in ``test/REQUIRE``.

   ``test(pkgs...)``

    Run the tests for each package in ``pkgs`` ensuring that each package's test dependencies are installed for the duration of the test. A package is tested by running its ``test/runtests.jl`` file and test dependencies are specified in ``test/REQUIRE``.
```

**Comment by jiahao** [comment]
A question for later design, maybe: I was discussing further with @jakebolewski the idea we floated offline about a second pass to detect cross-referrable entities within the docs. @jakebolewski reminded me that the help database currently isn't reified until the first help query is run. Does this mean that cross-references can only be generated after help is called once? Or will there need to be a second pass to generate rendered docstrings with cross-references statically? If so, how will runtime extensibility from loading new modules work?

**Comment by StefanKarpinski** [comment]
@one-more-minute, I'm glad you're taking another crack at this even after I got you into hot water the last time around.

**Comment by one-more-minute** [comment]
@StefanKarpinski No problem! Looking forward to this being done and dusted though.

@jiahao So to try and make it clear what's going on here: In the previous system, the *.rst files were the primary source for docstrings, and they were loaded into Julia as a second step. This PR flips that around – [docs/helpdb.jl](https://github.com/JuliaLang/julia/blob/66a5b0f2545cb3aa23ccb7d9a44e5d17c8221d56/base/docs/helpdb.jl#L14674-L14685) is now the primary source for docstrings like `Pkg.test()`, and those docs are spliced into the manual as a second step.

That means that once this is in, the old help system will be entirely redundant – its mechanics aren't really relevant to us here.

Things like cross references are worth discussing – I think we can make them a lot nicer than they were before, and it can all be done in a dynamic way. However, since this flip-around step is (hopefully) the most painful part of the process, I think it makes sense to make it as simple as possible, focus on moving without information loss, and build on it from there – deleting the help system, translating to markdown, implementing cross-references etc. etc., will be much easier if we can build them on top of the reorganised system.

It's a little unfortunate that the docs have to get worse before they get better, and obviously I'd love to do this stuff all in one go, but I just don't see a route that involves a reasonable level of effort. I think the very short-term pain will be well worth it long run.

(Although arguably, this actually improves the REPL experience, since we can view more than one method of the function now.)

**Comment by jiahao** [comment]
Thanks Mike, that explanation was very helpful.

I've examined the troublesome cases from the previous incarnation of this PR and they all look much better than before. Separating the issue of migrating from manual to docstring from ReST conversion is much clearer and avoids all the formatting weirdness like changing backquoting to double quotes.

There is an issue with docstrings shared across multiple functions, e.g. [`!=` and `≠`](https://github.com/JuliaLang/julia/blob/66a5b0f2545cb3aa23ccb7d9a44e5d17c8221d56/base/docs/helpdb.jl#L11425). If I understand correctly, the new helpdb.jl only associates the existing docstring to `Base.(:(!=))` but not `Base.(:(≠))`. The easiest thing to do might be to duplicate the existing docstring, but maybe it's worth 5 minutes of thought on how to handle share docstrings.

There's also a minor formatting problem in the docstring itself, where only the first method is double-backquoted but other methods are not. That can be fixed easily though.

**Comment by one-more-minute** [comment]
Ok great, thanks for the feedback. Would it be sufficient to indent all signatures, i.e. changing that to
```
```rst
           !=(x, y)
           ≠(x,y)
...
```
for them all to be properly formatted as code?

We can handle shared docstrings reasonably nicely already by doing:
```julia
@doc (@doc foo) bar
```
It might be useful, and would be reasonably easy to implement, to have proper aliases for metadata. It might make sense to handle these cases as part of this translation – I'll play around with it.

**Comment by ScottPJones** [comment]
Since I think a lot of cases of Unicode operators have (or should have) an ASCII equalivalent operator or function name, it does seem like it would be good to make that sharing handled.
:+1: for idea of having proper aliases for the metadata.

**Comment by jiahao** [comment]
Indentation would suffice in ReST when prefaced with a line containing only "::"

**Comment by one-more-minute** [comment]
Actually, unicode characters like `≠` are just aliases for the original function anyway, as opposed to functions in their own right. So we don't need to do anything more to handle that properly.

**Comment by ScottPJones** [comment]
I'm pretty sure I've seen cases where there was just a Unicode character, and no ASCII alias

**Comment by jiahao** [comment]
The most complicated example of shared docstrings is `in`, `∋`, `∈`, `∉`: https://github.com/JuliaLang/julia/commit/66a5b0f2545cb3aa23ccb7d9a44e5d17c8221d56#diff-badd7d54b8cfdba53f7e1064089334adR160

Not all of these are exact synonyms. Can this case be handled? (If not, one serious alternative is to document only `in`, `∋` and re-document the rest.)

File: doc/stdlib/arrays.rst [hunk 1]
```diff
@@ -176,12 +240,48 @@ Constructors
...
+
+   ::
+              eye(m, n)
+
    m-by-n identity matrix
 
+   ::
+              eye(A)
+
+   Constructs an identity matrix of the same dimensions and type as ``A``.
```
Comment (hayd) [inline doc/stdlib/arrays.rst]: `eye(A)` is repeated 3 times.

File: doc/stdlib/base.rst [hunk 2]
```diff
@@ -62,8 +75,21 @@ Getting Around
 
    Show a file using the default pager, optionally providing a starting line number. Returns to the julia prompt when you quit the pager.
 
+   ::
+              less(function, [types])
```
Comment (hayd) [inline doc/stdlib/base.rst]: same for `less(function, [types])` and `edit(function, [types])`

File: doc/stdlib/base.rst [hunk 3]
```diff
@@ -368,13 +429,26 @@ Types
...
 
+   ::
+              sizeof(s::AbstractString)
+
+   The number of bytes in string ``s``.
+
 .. function:: eps([type])
 
-   The distance between 1.0 and the next larger representable floating-point value of ``type``. Only floating-point types are sensible arguments. If ``type`` is omitted, then ``eps(Float64)`` is returned.
+   ::
```
Comment (hayd) [inline doc/stdlib/base.rst]: has the docstring for `eps([type])` has been replaced with `eps(::DateTime)` ?

File: doc/stdlib/arrays.rst (hunk 1 above)
Comment (one-more-minute) [inline doc/stdlib/arrays.rst]: Yes, this is the methods issue mentioned in the PR description – three methods are listed in the manual, the methods are combined for the whole function, each method in the manual gets the docs for the whole function.

File: doc/stdlib/base.rst (hunk 3 above)
Comment (one-more-minute) [inline doc/stdlib/base.rst]: Yes, also mentioned in the PR description.

**Comment by one-more-minute** [comment]
@jiahao Absolutely, we can handle those cases easily enough. Although if they're not actually synonyms it probably isn't a bad idea to give them separate docstrings with reference to each other.

**Comment by one-more-minute** [comment]
How does this look now? I've edited the head of the PR to look at the issues that aren't addressed in the code itself, but let me know if I've missed any feedback.

File: doc/stdlib/math.rst [hunk 4]
```diff
@@ -788,13 +907,11 @@ Mathematical Functions
 
 .. function:: trunc([T,] x, [digits, [base]])
 
-   ``trunc(x)`` returns the nearest integral value of the same type as ``x`` whose absolute
```
Comment (tkelman) [inline doc/stdlib/math.rst]: two methods for `trunc` getting clobbered by the `trunc` method in `Dates` ?

File: doc/stdlib/libc.rst [hunk 5]
```diff
@@ -40,7 +40,10 @@
 
 .. function:: time(t::TmStruct)
 
-   Converts a ``TmStruct`` struct to a number of seconds since the epoch.
```
Comment (tkelman) [inline doc/stdlib/libc.rst]: lost this line?

File: doc/stdlib/math.rst (hunk 4 above)
Comment (one-more-minute) [inline doc/stdlib/math.rst]: Yes, that looks like it's the same issue as with `eps`.

File: doc/stdlib/libc.rst (hunk 5 above)
Comment (one-more-minute) [inline doc/stdlib/libc.rst]: That's due to clobbering by `Dates` again, same issue as above. It's not [lost in helpdb](https://github.com/JuliaLang/julia/blob/c742260b589799c5ae42e61b8606d6f67603479c/base/docs/helpdb.jl#L580-L588) so can be recovered.

Comment (one-more-minute) [inline doc/stdlib/libc.rst]: Wait, hmm, the Libc.doc isn't actually right – better look into that

Comment (tkelman) [inline doc/stdlib/libc.rst]: getting really close here that these are the only losses of data - duplication isn't as much of an issue I think, unless sphinx chokes on things

Comment (one-more-minute) [inline doc/stdlib/libc.rst]: Yes, I think we're really close now – thanks for the eagle-eyed review!

Comment (one-more-minute) [inline doc/stdlib/libc.rst]: Ok, ignore me, I misread the `Libc.time` docstring and it's actually OK – this is just the method-clobbering issue again but the data isn't lost.

Comment (StefanKarpinski) [inline doc/stdlib/libc.rst]: It's hard to follow this conversation purely over email, but what about
doing docstrings for multiple things by applying the docstring to a literal
array of objects? E.g.
>
>
"doc for `foo`, `bar` and `baz`"
[foo, bar, baz]


I can't see any reason one would ever need to actually attach documentation
to a literal array that isn't assigned anywhere, so this seems like it
would be a safe syntax to interpret specially. It might be nice to document
constant arrays on the other hand so that you can, e.g. write a doc string
for LOAD_PATH and then do things like

help?> LOAD_PATH


and get the docstring associated with that array object. Although perhaps
that should be done at the global symbol level instead of the object level.

Comment (one-more-minute) [inline doc/stdlib/libc.rst]: Yes, that seems like a reasonable way to share docstrings. I definitely agree that we need a way to document bindings themselves as opposed to just objects – that's something I have an implementation in mind for and I just need some time to play around with it.

Comment (MichaelHatherly) [inline doc/stdlib/libc.rst]: > a literal array of objects?

That syntax has worked quite nicely in Docile &ndash; especially for documenting functions with mutating and non-mutating versions:

```jl
"..."
[foo, foo!]
```

---

@one-more-minute would it be possible to split ``base/docs/helpdb.jl`` into several files, perhaps with the same structure as ``docs/stdlib/*``? Might help highlight anything else that's missing/duplicated more clearly.

Comment (one-more-minute) [inline doc/stdlib/libc.rst]: Splitting things up is a good idea, I can give that a go.

I think that if we're seriously considering adding more syntaxes like this, #11850 becomes much more important.

**Comment by JeffBezanson** [comment]
What's the status of this? Anything I or anyone else can do to help? Let's get an RC out!

**Comment by one-more-minute** [comment]
As long as there are no more major info losses that we can deal with systematically, I think this should basically be good to go (modulo a rebase). Once it's in I can work on what can be converted automatically, and we can crowdsource moving things around and fixing the remaining issues.

I didn't get round to doing @MichaelHatherly's idea of splitting up the huge `helpdb.jl` file, can have a go at that if it's wanted.

**Comment by JeffBezanson** [comment]
+1 for getting this merged and crowdsourcing the needed doc touch-ups.

**Comment by johnmyleswhite** [comment]
+1 to that

**Comment by one-more-minute** [comment]
Ok, I rebased, assuming there's no other major objections we should be good to go. Speak now or forever hold your peace etc.

**Comment by yuyichao** [comment]
Seems that both AppVeyor and Travis have something to say... ;-p

**Comment by one-more-minute** [comment]
Think I might just have been unlucky there – looks like networking issues?

**Comment by yuyichao** [comment]
Yeah, looks like the win64 freeze plus github network issues... Restarted.

File: doc/stdlib/strings.rst [hunk 6]
```diff
@@ -377,31 +554,56 @@
...
-.. function:: utf16(::Union{Ptr{UInt16},Ptr{Int16}} [, length])
+   ::
+              utf16(::Union{Ptr{UInt16},Ptr{Int16}} [, length])
 
    Create a string from the address of a NUL-terminated UTF-16 string. A copy is made; the pointer can be safely freed. If ``length`` is specified, the string does not have to be NUL-terminated.
 
-.. function:: utf32(s)
+.. function:: utf16(::Union{Ptr{UInt16},Ptr{Int16}} [, length])
 
-   Create a UTF-32 string from a byte array, array of ``Char`` or ``UInt32``, or
```
Comment (tkelman) [inline doc/stdlib/strings.rst]: lost this one?

**Comment by tkelman** [comment]
There are still the dates module conflicts in a few places. If the rst docs are now auto-generated from the helpdb.jl, touching it up manually won't do much good if it gets automatically overwritten. Should this be deleting the existing copy of helpdb.jl if it's no longer being used?

**Comment by one-more-minute** [comment]
Yes, the issues you point out are due to the names being aliases to a single function, as I've mentioned. The information is preserved correctly in helpdb as you can verify for yourself.

I could delete the old helpdb in this PR, but the idea is to make it as simple as possible so it's easy to review and rebase. I don't think it's hurting anyone while it's there.

**Comment by tkelman** [comment]
Well, exactly how does this work then, and what are the goals for how much more will be done prior to vs after 0.4? Will sphinx and the old helpdb go away before then? If the new helpdb contains more information, why couldn't we switch to using it as the source of repl help right away? It doesn't look like there's enough of a change within base for that to be the case yet?

If we'll still be using sphinx and rst somewhere in the doc pipeline for the 0.4.0 release, then I think the last module-clash problems here should be fixed in the automated scripts otherwise information will get lost every time we re-run the docstring->rst translation.

**Comment by JeffBezanson** [comment]
> I could delete the old helpdb

I'd strongly favor simply deleting everything related to the old help system (assuming it's unused, of course). Having deprecated code lying around is confusing. Doesn't have to be done as part of this PR, but should be done very soon.

What should I do if I want to write or update function help today? Is it safe to keep updating the existing .rst files?

**Comment by one-more-minute** [comment]
Rebased and borrowed Stefan's helpdb removal commit.

The manual is safe to edit apart from docstrings themselves (signatures are fine too). To edit docstrings we'll be editing them right at the source from now on.

**Comment by StefanKarpinski** [comment]
When you get done with this, can you make an announcement explaining the transition, where people now need to edit stdlib documentation, and asking for help with the rest of the transition? I suspect we can get it done in fairly short order if it's easy to help.

**Comment by one-more-minute** [comment]
Sure, happy to make an announcement. I think that before we start crowdsourcing it would be good to translate as much as possible automatically, so I'll announce the changes, then do that, then call for help.

**Comment by StefanKarpinski** [comment]
Sounds like a plan!

File: CONTRIBUTING.md [hunk 7]
```diff
@@ -152,7 +152,7 @@ Make sure that [Travis](http://www.travis-ci.org) greenlights the pull request w
  - Try to [squash](http://gitready.com/advanced/2009/02/10/squashing-commits-with-rebase.html) together small commits that make repeated changes to the same section of code so your pull request is easier to review, and Julia's history won't have any broken intermediate commits. A reasonable number of separate well-factored commits is fine, especially for larger changes.
  - If any conflicts arise due to changes in Julia's `master`, prefer updating your pull request branch with `git rebase` versus `git merge` or `git pull`, since the latter will introduce merge commits that clutter the git history with noise that makes your changes more difficult to review.
  - If you see any unrelated changes to submodules like `deps/libuv`, `deps/openlibm`, etc., try running `git submodule update` first.
- - Avoid committing changes to auto-generated files such as `doc/helpdb.jl`, which is a frequent source of conflicts and can be re-generated later.
+ - Avoid committing changes to auto-generated files such as `doc/stdlib/*.rst`, which is a frequent source of conflicts and can be re-generated later.
```
Comment (hayd) [inline CONTRIBUTING.md]: I wonder if it would be more obvious that these were generated files (and so shouldn't be changed) if it was `/_stdlib/*.rst`.

**Comment by one-more-minute** [comment]
Not sure what's going on with the build, but the makefile changes may have broken it.

File: test/docs.jl [hunk 8]
```diff
@@ -21,206 +21,3 @@ end
 
 @test (@doc ModuleMacroDoc)    == "I am a module"
 @test (@doc ModuleMacroDoc.@m) == ["I am a macro"]
-
-# apropos function testing
```
Comment (tkelman) [inline test/docs.jl]: what's with the removal of tests?

**Comment by tkelman** [comment]
There are a few mentions of helpdb in makefile dependencies that you'll need to adjust. The automated doc steps should be added to the `release-candidate` target too.

Has any status changed on the module conflicts with Dates, and `utf32` which seems to have been lost? Manual touchups that would get clobbered the next time the automated process runs doesn't seem like a valid solution.

File: test/docs.jl (hunk 8 above)
Comment (one-more-minute) [inline test/docs.jl]: If you're removing deprecated code you sorta have to get rid of the tests for that code, too.

Comment (yuyichao) [inline test/docs.jl]: Incorrect resolution of rebase conflict I guess?

Comment (tkelman) [inline test/docs.jl]: Sorry, missed that this is deleting not just the old copy of helpdb but also the old help system that was using it. Seems like a fair amount of the tests here should still be valid though?

**Comment by JeffBezanson** [comment]
Bump. This is on the critical path for 0.4. Status update?

**Comment by one-more-minute** [comment]
I rebased and patched the helpdb removal – looks like the tests are ok now.

This is good to go as far as I'm concerned. If we can get it in then I can do the announcements and get the markdown conversion going ASAP tomorrow.

**Comment by StefanKarpinski** [comment]
Sounds good to me. @jiahao, are you good with this?

File: Makefile [hunk 9]
```diff
@@ -96,7 +96,6 @@ release-candidate: release testall
 	@$(MAKE) -C doc latex SPHINXOPTS="-n" #Rebuild Julia PDF docs pedantically
 	@$(MAKE) -C doc doctest #Run Julia doctests
 	@$(MAKE) -C doc linkcheck #Check all links
-	@$(MAKE) -C doc helpdb.jl #Rebuild Julia online documentation for help(), apropos(), etc...
```
Comment (tkelman) [inline Makefile]: please add the new doc building steps here instead

File: test/docs.jl [hunk 10]
```diff
@@ -22,73 +22,6 @@ end
 @test (@doc ModuleMacroDoc)    == "I am a module"
 @test (@doc ModuleMacroDoc.@m) == ["I am a macro"]
 
-# apropos function testing
-
-@test sprint(apropos, "non-documented object") == "No help information found.\n"
```
Comment (tkelman) [inline test/docs.jl]: is `apropos` completely gone now?

**Comment by tkelman** [comment]
Still unaddressed:

> Has any status changed on the module conflicts with Dates, and utf32 which seems to have been lost? Manual touchups that would get clobbered the next time the automated process runs doesn't seem like a valid solution.

**Comment by one-more-minute** [comment]
> Will sphinx and the old helpdb go away before then?

I'm no authority on what should go into 0.4, but sure, I'm happy to push towards the new manual this week or whenever.

> I think the last module-clash problems here should be fixed in the automated scripts

You can think that all you want, but to actually do it I need a bit more of a concrete idea how that might work. Right now it doesn't seem to me that we can do it in an automated way, which is why it makes sense to crowdsource it.

> information will get lost every time we re-run the docstring->rst translation.

Why?

> Has any status changed on the module conflicts with Dates, and utf32 which seems to have been lost?

They haven't been lost, they're in helpdb and simply aren't presented in the manual right now. The way to handle this is to add type information to the method signatures in both helpdb and the manual so that information is stored and retrieved at the method level, rather than at the function level.

> Manual touchups that would get clobbered the next time the automated process runs doesn't seem like a valid solution.

Pretty sure no one has proposed this as a valid solution.

**Comment by tkelman** [comment]
Okay, then you really need to explain the architecture of exactly how this works and the short-term goals in a lot more detail than just https://github.com/JuliaLang/julia/pull/11943#issuecomment-117219742

And everyone else needs to decide if we want to keep the current rst docs in place for 0.4 or not. If not, then there's a whole lot of work that remains to be done to follow up on this to get an acceptable replacement in place very quickly, and it's not clear that anyone other than you has any idea what the steps to get there will be.

>> information will get lost every time we re-run the docstring->rst translation.
> Why?

Assuming the docstring->rst translation code doesn't get fixed, if we keep using rst and sphinx then we'll need to re-run that code every time we want to update the rendered html docs.

> The way to handle this is to add type information to the method signatures in both helpdb and the manual so that information is stored and retrieved at the method level, rather than at the function level.

Wasn't exactly this recommended several times before by @nalimilan and others? How hard is this to implement?

There seems to be a contradiction here between

> Right now it doesn't seem to me that we can do it in an automated way, which is why it makes sense to crowdsource it.

and

> Pretty sure no one has proposed this as a valid solution.

What are you proposing to crowdsource then, if not manual touchups to rectify gaps in auto-generated content?

**Comment by one-more-minute** [comment]
If people aren't sure what's happening or what the plan is then I'm more than happy to answer any questions, which is a little more constructive than making unfounded assumptions about what's wrong with it or demands about how it should be done.

> if we keep using rst and sphinx then we'll need to re-run that code every time we want to update the rendered html docs.

Yes, of course. But you said that information will be lost every time we do this – why?

> Wasn't exactly this recommended several times before by @nalimilan and others?

Ok, well we'll make sure I don't get undue credit for that idea if that concerns you.

> What are you proposing to crowdsource then, if not manual touchups to rectify gaps in auto-generated content?

I am proposing manual touchups, just not to auto-generated content that will get overwritten – one might credit me a little more competence than that. As I did say before, "The manual is safe to edit apart from docstrings themselves (signatures are fine too)." The signatures can be edited without being overwritten, and once they have type info that can be used as part of the manual-generation step.

**Comment by tkelman** [comment]
I'm not trying to come across as a dick here, please don't take my criticism personally - most of the questions I'm asking are out of ignorance about how this works and what the goals are. And I still don't think you've described this thoroughly enough yet. This is a failure of communication, please explain more (a *lot* more) than you think is necessary here, we can't read your mind and major changes to the documentation really can't be done lightly.

> But you said that information will be lost every time we do this – why?

Touchups to autogenerated content would be lost. If the code that does this autogeneration doesn't get any smarter, than information will be lost between current rst docs and future autogenerated rst docs. The question is whether or not that matters for 0.4.

> I am proposing manual touchups, just not to auto-generated content that will get overwritten – one might credit me a little more competence than that.

You haven't been explaining your thought process anywhere near as clearly as you think you have, please be more detailed or people will continue to misunderstand what's happening here.

File: test/docs.jl (hunk 10 above)
Comment (tkelman) [inline test/docs.jl]: Looks like there are still vestiges of it at
```
base/docs/basedocs.jl:20:  You can also use `apropos("...")` to search the documentation.
base/exports.jl:1077:    apropos,
contrib/BBEditTextWrangler-julia.plist:192:              <string>apropos</string>
doc/stdlib/base.rst:144:.. function:: apropos(string)
```
which should be deleted if it's being completely removed

**Comment by ViralBShah** [comment]
It would certainly help me to know what remains to be done, perhaps have a checklist, and also how to try out the new documentation. It seems like a lot of it is ready, but some instructions would help a long way.

This is the last major thing for 0.4, so I am happy to help out in any way I can - I just don't know what to do yet, since this was a difficult transition, and it has been a bit difficult to follow all that has been going on in this space.

**Comment by one-more-minute** [comment]
I'll try to lay out the steps more thoroughly. Likewise, though, I can't read people's minds to figure out where misunderstandings are – I'm happy to clarify any of the steps in more detail if they don't seem comprehensive enough.

The initial steps are:

1. Moving the docstrings from the manual into the Base source code. This means that docstrings are "live" in Julia's memory, can be accessed from the REPL etc. It also makes the old help system redundant, since the new doc system prioritises itself over information in helpdb.
  * Since the manual contains docs for individual methods, but no way of automatically identifying those methods in general (i.e. type information) a somewhat shotgun-ish approach is necessary here. Method-level documentation is concatenated and applied to the function as a whole. This has its caveats but actually improves the REPL help since more than one method is now shown per function.
2. Making the Base docstrings the primary source of function documentation. This is necessary in order to do the conversion to Markdown and make the Base docstrings really first-class within the system. Clearly, the manual must be generated from the docstrings as part of this. The way this works is that the `genstdlib.jl` script reaches into the manual, looks for docstrings, identifies the appropriate object and its metadata, and replaces the *content* of the rst docstring with that metadata. As you can tell from the diff, only the content of docstrings is affected while the main-body text and function signatures are left alone, which means that they can be freely edited without being clobbered later on.
  * Caveat: the issue above means there's no method-level fidelity. Multiple methods in the manual simply repeat the whole function docs multiple times, and the same function can't be adequately documented from multiple modules.
3. Convert to Markdown. While the docstrings are all in one file, it's easy to pull the RST out of that file, feed it to pandoc, and splice Markdown back in. This will work for nearly 90% of the docstrings with the rest sorted out below.

Following that, the steps (in no particular order) are:

1. Adding type information. Since this simply doesn't exist at this stage it must be added by hand. Docstrings can be associated with methods via (3), the manual will have a simple syntax for specifying a method, and `genstdlib.jl` can hook them together in the right way easily enough. The sub-tasks here are to get the method-level stuff working, produce an example PR adding signatures to point people in the right direction, and call for help to crowdsource the bulk of the work.
2. Converting the remaining docstrings to MD. This will require an implementation of things like cross references, as well as a way to convert those to RST if we still have the old manual at this stage. As above, the steps are to get this working, produce an example, and crowdsource.
3. Moving the docstrings to the relevant code in Base, so that the docs are inline.
4. Moving the manual itself to Markdown. There are still some key decisions to be made here, and work to be done generating decent, searchable HTML and host it.

These tasks are fairly daunting, but very easy to parallelize, and none of them is particularly dependent on the other – perfect for some crowdsourcing goodness.

**Comment by pao** [comment]
5, leave stub ReadTheDocs-compatible documentation in place which directs people to the new site that doesn't exist yet.

**Comment by tkelman** [comment]
Great, thanks, that's about the right level of impossible-to-misunderstand explanation for a change this big.

To make sure I've got this right - moving the docstrings for the handful of small clashes / lost-rst-data to the methods in question will automatically populate type info from the method signature, then even staying in rst that data could be brought back? Is that type-information population from methods currently working when the docstrings are in the right place, or does the whole system for doing that need to be written?

Personally it sounds like there's too much work and too many unknowns to actually get all the way to having everything in markdown to be worth making that an 0.4 target. My vote is to find a solution to restore the rst data lost from module conflicts here, use rst for RC1, and cautiously see how far markdown conversion can get during the RC stage. Might be better to have the totally new doc website be built as a work in progress on master for 0.5, rather than released with 0.4 while still brand new.

Small things then I think this PR can be merged:
- If `apropos` is going away, then get rid of all signs of it
- `genstdlib.jl` should be run during `make release-candidate`

And immediately after merging:
- All open PR's that touch docs will need to be notified and rebased to move their doc modifications to the helpdb docstrings

**Comment by ViralBShah** [comment]
I think @tkelman has the right idea. Perhaps we can start the RC process and see how far this goes. 

Another thought is whether some middle-path possible for 0.4, where we convert from RST to Markdown and start using the new docsystem, and do the full transition of the documentation into Base sources in 0.5?

File: doc/devdocs/cartesian.rst [hunk 11]
```diff
@@ -200,6 +204,7 @@ Macro reference
     would generate::
 
         if i_1 > size(A, 1)
+
```
Comment (tkelman) [inline doc/devdocs/cartesian.rst]: Odd. Bug maybe?

**Comment by StefanKarpinski** [comment]
Ok, we're doing this and going to deal with various issues that need to be fixed as they come up.

**Comment by aviks** [comment]
Woo hoo! Thanks @one-more-minute, @MichaelHatherly  et al...

**Comment by tkelman** [comment]
Yeah, probably the right thing to do. Would still be nice to get answers to the honest questions I had above. Hard to fix known issues if no one else knows how to go about fixing them. And we'll have to be very careful about open PR's that touch parts of the docs that are now autogenerated from docstrings.

**Comment by StefanKarpinski** [comment]
I agree, but we needed to move on this and get a 0.4 RC out the door.

**Comment by tkelman** [comment]
Sure, my 2 small "before merging" bullet items were taken care of. Now we all need to clean this up and let everyone who has an open PR or will file any future PRs touching docs know what they need to do. I still don't get where or what needs to be done to restore the data that's currently lost from the rst docs, do you?

As far as @ViralBShah's idea above goes, I was actually thinking the other way around. This PR created the docstrings already so now we should hopefully be free to move them around however we want, but conversion from rst to markdown and moving away from sphinx seems way premature and not something we should try to do while we're worrying about release candidates.

**Comment by tkelman** [comment]
All open PR's that touch docs (likely an overestimate):

#12432
#12429
#12423
#12408
#12385
#12383
#12353
#12350
#12333
#12325
#12146
#12083
#12042
#12033
#12025
#11998
#11977
#11842
#11754
#11631
#11242
#11077
#11012
#10940
#10810
#10800
#10799
#10776
#10760
#10612
#10517
#10369
#10277
#10108
#9798
#9596
#9340
#8775
#8685
#8668
#8641
#8636
#8258
#7327
#7186
#6842
#6669
#6455

**Comment by KristofferC** [comment]
So what's the next step now? I read the comments but I'm still not sure. Is it to transfer the stuff from `helpdb.jl` to the source files?

**Comment by ScottPJones** [comment]
:+1: :100: for doing this!  Now to iron out any glitches before 0.4.0 release.
I just got a build with new master, and found that the `?` is not giving any information for some functions that it used to (I was hoping that my document strings in Base would be picked up now).
```julia
help?> convert(UTF32String, Char['a','b','c'])

help?> convert(UTF8String, Char['a','b','c'])

help?> convert(UTF8String, UInt8[64,65,66])

help?> convert(Int, 3.0)
```
Old:
```julia
help?> convert(Int, 3.0)
convert(::Type{Int64}, x::Float64) at int.jl:205

julia> convert(UTF8String, UInt8[64, 65, 66])
"@AB"

help?> convert(UTF8String, UInt8[64, 65, 66])
convert(::Type{UTF8String}, dat::Array{UInt8,1}) at unicode/utf8.jl:231
```

**Comment by tkelman** [comment]
That looks like a regression that is probably worth a separate issue.

**Comment by amitmurthy** [comment]
Echoing @KristofferC , what is the next step for now? If changes are not expected to be made to `doc/stdlib/*.rst` and the source files do not yet carry documentation, do we just hold off on stdlib doc changes for now?

**Comment by one-more-minute** [comment]
Docstring changes are fine, but they should be made to `helpdb.jl` as opposed to `*.rst`.

The next step is to convert stuff to markdown, #12435, and once that's in we can all start to move the docstrings to appropriate places.

**Comment by tkelman** [comment]
Don't know if you missed this or are ignoring me, but again, would be great to get answers for the following

> moving the docstrings for the handful of small clashes / lost-rst-data to the methods in question will automatically populate type info from the method signature, then even staying in rst that data could be brought back? Is that type-information population from methods currently working when the docstrings are in the right place, or does the whole system for doing that need to be written?

**Comment by one-more-minute** [comment]
Sure, have just had a lot on so things might've slipped by. Just moving the docstrings above the correct methods will give them the correct type info, yes – all that stuff's already in place. Then we just need a syntax for type info in the RST docs and for genstdlib.jl to recognise that syntax, e.g. we'd change

```
.. function:: length(a)
```
to
```
.. function:: length(a::AbstractArray)
```
so that the specific method doc is spliced in instead of the whole function.

**Comment by tkelman** [comment]
Awesome. I'll experiment with moving around the functions getting clobbered by Dates module docs, and `utf32` which currently aren't making it into RST correctly and see if it helps bring those back. (Though maybe you'll want to get #12435 done first, so I'll do it in a PR.)

Some of the RST docs have already been using declaration-like type syntax, but not all. `genstdlib.jl` tweaks will probably need your help.

**Comment by one-more-minute** [comment]
Ok, great – it also occurs to me that I can actually just run the #12435 conversion over all of Base if necessary, so it's not really a big deal if we start moving docstrings before then.

**Comment by tkelman** [comment]
Well, tried https://gist.github.com/b5c7657d53d5c37b4e85 but so far that doesn't appear to be helping much. And at the repl this might just be #12437?

```
help?> eps
search: eps RepString @elapsed indexpids expanduser escape_string peakflops unescape_string

  Returns Millisecond(1) for DateTime values and Day(1) for Date values.

help?> Base.eps
  Returns Millisecond(1) for DateTime values and Day(1) for Date values.

help?> eps(Date(1))

help?> eps()

help?> eps(1.0)
  Returns Millisecond(1) for DateTime values and Day(1) for Date values.
```

I think the top priority needs to be getting the repl help usable and fixing the manual back up, more than worrying about what format it's in.

File: CONTRIBUTING.md (hunk 7 above)
Comment (tkelman) [inline CONTRIBUTING.md]: I think we should actually delete this line, since the rst files are only partially auto-generated. Best to just commit both copies of the docstrings for now I think.
